tools/normalize.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2020 Codethink Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <inttypes.h>
+#include <string.h>
+
+#include "../src/msg.h"
+
+#include "msg.h"
+#include "sig.h"
+#include "fifo.h"
+
+// DEFAULT_AVERAGE_WIDTH, the number of milliseconds to average over
+long DEFAULT_AVERAGE_WIDTH = 1000;
+
+struct channel_data {
+	bl_msg_sample_data_t msg; // Used as a buffer and for outputting the processed samples
+	struct fifo samples;      // Store samples until we've finished using them for processing
+	double rolling_average;   // The rolling average as of the most recent sample
+};
+
+void add_normalized_sample(struct channel_data *channel, uint32_t sample) {
+	// If adding it to the message is impossible because it's full, write the message.
+	// sample may be 16-bit or 32-bit full of numbers, depending on data format.
+	if (channel->msg.type == BL_MSG_SAMPLE_DATA16) {
+		channel->msg.data16[channel->msg.count] = (uint16_t) sample;
+		channel->msg.count++;
+	} else {
+		channel->msg.data32[channel->msg.count] = sample;
+		channel->msg.count++;
+	}
+	if ((channel->msg.type == BL_MSG_SAMPLE_DATA16 && channel->msg.count == MSG_SAMPLE_DATA16_MAX)
+	    || (channel->msg.type == BL_MSG_SAMPLE_DATA32 && channel->msg.count == MSG_SAMPLE_DATA32_MAX)) {
+			// Message is full, send and start refilling
+			bl_msg_yaml_print(stdout, (union bl_msg_data *) &channel->msg);
+			channel->msg.count = 0;
+	}
+}
+
+int add_sample(long average_width, struct channel_data *channel, uint32_t sample)
+{
+	// Add a sample to the fifo
+	if (!fifo_write(&channel->samples, sample)) {
+		fprintf(stderr, "FIFO overflow\n");
+		return -ERANGE;
+	}
+	// Add to the rolling average
+	channel->rolling_average += (float) sample / (float) average_width;
+	if (channel->samples.used >= average_width) {
+		uint32_t new_sample, old_sample;
+		double normalized_sample, sample_ratio;
+
+		// Calculate the normalized sample
+	   	if (!fifo_peek_back(&channel->samples, average_width / 2, &new_sample)) {
+			fprintf(stderr, "FIFO underflow\n");
+			return -ERANGE;
+		}
+		sample_ratio = (float) new_sample / channel->rolling_average;
```

Comment:
With rolling sum, something like:

```suggestion
		normalized_sample = new_sample * average_width * channel->baseline / channel->rolling_sum;
```

Where `channel->baseline` is a `uint32_t` containing either `INT16_MAX` or `INT32_MAX` depending on the channel sample type.

With average_width passed in as a `uint64_t`.  In fact, I'd be tempted to rename "average_width" to something like "window_size", which is more idiomatic for signal processing I think.

* All completely untested!  :smile_cat: